app/plugins/serverless-functions.md

```diff
@@ -0,0 +1,157 @@
+  consumer_id: false
+  config:
+    - name: functions
+      required: true
+      default: "[]"
+      value_in_examples: "[]"
+      description: Array of stringified Lua code to be cached and run in sequence during access phase.
+
+---
+
+> **Note**: Serverless plugins are available in Enterprise as of version 0.32 
+> and will soon be available as part of Kong CE 0.14.
+
+## Plugin Names
+
+Serverless Functions come as two separate plugins. Each one runs with a
+different priority in the plugin chain.
+
+- `pre-function`
+  - Runs before other plugins run during access phase.
+- `post-function`
+  - Runs after other plugins in the access phase.
+
+## Demonstration
+
+1. Create a Service on Kong:
+
+    ```bash
+    $ curl -i -X  POST http://localhost:8001/services/ \
+      --data "name=plugin-testing" \
+      --data "url=http://httpbin.org/headers"
+
+    HTTP/1.1 201 Created
+    ...
+    ```
+
+2. Add a Route to the Service:
+
+    ```bash
+    $ curl -i -X  POST http://localhost:8001/services/plugin-testing/routes \
+      --data "paths[]=/test"
+
+    HTTP/1.1 201 Created
+    ...
+    ```
+    
+1. Create a file named `custom-auth.lua` with the following content:
+
+    ```lua
+    -- Get list of request headers
+    local custom_auth = kong.request.get_header("x-custom-auth")
+
+    -- Terminate request early if our custom authentication header
+    -- does not exist
+    if not custom_auth then
+      return kong.response.exit(401, "Invalid Credentials")
+    end
+
+    -- Remove custom authentication header from request
+    kong.request.clear_header('x-custom-auth')
+    ```
+
+4. Ensure the file contents:
+
+    ```bash
+    $ cat custom-auth.lua
+    ```
+
+5. Apply our Lua code using the `pre-function` plugin using cURL file upload:
+
+    ```bash
+    $ curl -i -X POST http://localhost:8001/services/plugin-testing/plugins \
+        --data "name=pre-function" \
+        --data "config.functions[]=@custom-auth.lua"
+
+    HTTP/1.1 201 Created
+    ...
+    ```
+
+6. Test that our lua code will terminate the request when no header is passed:
+
+    ```bash
+    curl -i -X GET http://localhost:8000/test
+
+    HTTP/1.1 401 Unauthorized
+    ...
+    "Invalid Credentials"
+    ```
+
+7. Test the Lua code we just applied by making a valid request:
+
+    ```bash
+    curl -i -X GET http://localhost:8000/test \
+      --header "x-custom-auth=key"
+
+    HTTP/1.1 200 OK
+    ...
+    ```
+
+This is just a small demonstration of the power these plugins grant. We were 
+able to dynamically inject Lua code into the plugin access phase to dynamically 
+terminate, or transform the request without creating a custom plugin or 
+reloading / redeploying Kong.
+
+In short, serverless functions give you the full capabilities of a custom plugin
+in the access phase without ever redeploying / restarting Kong. 
+
+----
+
+### Notes
+
+#### Fake Upstreams
+
+Since the [Service][service-url] entity requires defining an upstream you may
```

Comment:
Same comment as before - I think this needs to be "must" instead of "may". If "may" is in fact correct, I suggest clarifying what happens if you don't have a fake upstream and terminate the request.

Reply:
It's `may`, in this context